nucypher/utilities/latency.py

```diff
@@ -0,0 +1,84 @@
+import time
+from collections import defaultdict
+from threading import Lock
+from typing import List
+
+from eth_typing import ChecksumAddress
+
+
+class NodeLatencyStatsCollector:
+    """
+    Thread-safe utility that tracks latency statistics related to P2P connections with other nodes.
+    """
+
+    CURRENT_AVERAGE = "current_avg"
+    COUNT = "count"
+    MAX_LATENCY = float(2**16)  # just need a large number for sorting
+
+    class NodeLatencyContextManager:
+        def __init__(
+            self,
+            stats_collector: "NodeLatencyStatsCollector",
+            staker_address: ChecksumAddress,
+        ):
+            self._stats_collector = stats_collector
+            self.staker_address = staker_address
+
+        def __enter__(self):
+            self.start_time = time.perf_counter()
+            return self
+
+        def __exit__(self, exc_type, exc_val, exc_tb):
+            if exc_type:
+                # exception occurred - reset stats since connectivity was compromised
+                self._stats_collector.reset_stats(self.staker_address)
+            else:
+                # no exception
+                end_time = time.perf_counter()
+                execution_time = end_time - self.start_time
+                self._stats_collector._update_stats(self.staker_address, execution_time)
+
+    def __init__(self):
+        # staker_address -> { "total_time": <float>, "count": <integer> }
+        self._node_stats = defaultdict(
+            lambda: {self.CURRENT_AVERAGE: 0.0, self.COUNT: 0}
+        )
+        self._lock = Lock()
+
+    def _update_stats(self, staking_address: ChecksumAddress, latest_time_taken: float):
+        with self._lock:
+            old_avg = self._node_stats[staking_address][self.CURRENT_AVERAGE]
+            old_count = self._node_stats[staking_address][self.COUNT]
+
+            updated_count = old_count + 1
+            updated_avg = ((old_avg * old_count) + latest_time_taken) / updated_count
```

Comment:
I think this is fine for this PR, but at a certain point, historical time measures are not relevant. I think a potential next iteration of this context manager can focus on collecting statistics for the last X hours or X usages. WDYT?

Reply:
Yep. We could do a moving average of `X` values eg. we could always store the last `X` values e.g. the last 5 latency times and do an ongoing average of that. Every new value would overwrite the oldest value.

Reply:
Fixed via https://github.com/nucypher/nucypher/pull/3562/commits/fb4152d8e20004d738f469c87bcfb12f971c88d7.

Used a window of 5. If you (or anyone else) have thoughts on a better value to use, let me know.